python/glow/wgr/linear_model/logistic_model.py

```diff
@@ -0,0 +1,306 @@
+
+    def fit(
+        self,
+        blockdf: DataFrame,
+        labeldf: pd.DataFrame,
+        sample_blocks: Dict[str, List[str]],
+        covdf: pd.DataFrame = pd.DataFrame({})
+    ) -> (DataFrame, DataFrame):
+        """
+        Fits a logistic regression model, represented by a Spark DataFrame containing coefficients for each of the ridge
+        alpha parameters, for each block in the starting matrix, for each label in the target labels, as well as a
+        Spark DataFrame containing the optimal ridge alpha value for each label.
+
+        Args:
+            blockdf : Spark DataFrame representing the beginning block matrix X
+            labeldf : Pandas DataFrame containing the target labels used in fitting the ridge models
+            sample_blocks : Dict containing a mapping of sample_block ID to a list of corresponding sample IDs
+            covdf : Pandas DataFrame containing covariates to be included in every model in the stacking
+                ensemble (optional).  The covariates should not include an explicit intercept term, as one will be
+                added automatically.  If empty, the intercept will be used as the only covariate.
+
+        Returns:
+            Two Spark DataFrames, one containing the model resulting from the fitting routine and one containing the
+            results of the cross validation procedure.
+        """
+        map_key_pattern = ['sample_block', 'label', 'alpha_name']
+        reduce_key_pattern = ['header_block', 'header', 'label', 'alpha_name']
+        model_key_pattern = ['sample_block', 'label', 'alpha_name']
+        score_key_pattern = ['sample_block', 'label']
+        metric = 'log_loss'
+
+        if not self.alphas:
+            self.alphas = generate_alphas(blockdf)
+
+        if covdf.empty:
+            covdf = pd.DataFrame(data=np.ones(labeldf.shape[0]),
+                                 columns=['intercept'],
+                                 index=labeldf.index)
+            validate_inputs(labeldf, pd.DataFrame({}), 'binary')
+        else:
+            covdf = covdf.copy()
+            validate_inputs(labeldf, covdf, 'binary')
+            covdf.insert(0, 'intercept', 1)
+
+        maskdf = pd.DataFrame(data=np.where(np.isnan(labeldf), False, True),
+                              columns=labeldf.columns,
+                              index=labeldf.index)
+
+        beta_cov_dict = {}
+        for label in labeldf:
+            row_mask = slice_label_rows(maskdf, label, list(labeldf.index), np.array([])).ravel()
+            cov_mat = slice_label_rows(covdf, 'all', list(labeldf.index), row_mask)
+            y = slice_label_rows(labeldf, label, list(labeldf.index), row_mask).ravel()
+            fit_result = constrained_logistic_fit(cov_mat,
+                                                  y,
+                                                  np.zeros(cov_mat.shape[1]),
+                                                  guess=np.array([]),
```

Comment:
We'll address this in a followup, but IMO it's better to use `None` as a dummy value than an empty array.